packages/near-membrane-dom/src/browser-realm.ts

```diff
@@ -166,10 +171,103 @@ export default function createVirtualEnvironment(
+        const oneTimeWindow = ReflectApply(
+            HTMLIFrameElementProtoContentWindowGetter,
+            oneTimeIframe,
+            []
+        )!;
+        defaultGlobalOwnKeys = getFilteredGlobalOwnKeys(oneTimeWindow);
+        ReflectApply(ElementProtoRemove, oneTimeIframe, []);
+        defaultGlobalPropertyDescriptorMap = {
+            __proto__: null,
+        } as unknown as PropertyDescriptorMap;
+        assignFilteredGlobalDescriptorsFromPropertyDescriptorMap(
+            defaultGlobalPropertyDescriptorMap,
+            ObjectGetOwnPropertyDescriptors(window)
+        );
+        for (let i = 0, { length } = defaultGlobalOwnKeys; i < length; i += 1) {
+            defaultGlobalOwnKeysRegistry[defaultGlobalOwnKeys[i]] = true;
+        }
+        for (const key in defaultGlobalPropertyDescriptorMap) {
+            if (!(key in defaultGlobalOwnKeysRegistry)) {
+                delete defaultGlobalPropertyDescriptorMap[key];
+            }
+        }
+    }
+    const blueRefs = getCachedGlobalObjectReferences(globalObjectVirtualizationTarget);
+    // Create a new environment.
+    const env = new VirtualEnvironment({
+        blueConnector,
+        distortionCallback,
+        redConnector: createConnector(
+            ReflectApply(FunctionProtoBind, ShadowRealmProtoEvaluate, [new ShadowRealmCtor()])
+        ),
+        instrumentation,
+    });
+    linkIntrinsics(env, globalObjectVirtualizationTarget);
+    // window
+    env.link('globalThis');
+    // Set globalThis.__proto__ in the sandbox to a proxy of
+    // globalObjectVirtualizationTarget.__proto__ and with this, the entire
+    // structure around window proto chain should be covered.
+    env.remapProto(globalObjectVirtualizationTarget, blueRefs.WindowProto);
+    let unsafeBlueDescMap: PropertyDescriptorMap = defaultGlobalPropertyDescriptorMap;
+    if (globalObjectVirtualizationTarget !== window) {
+        unsafeBlueDescMap = { __proto__: null } as unknown as PropertyDescriptorMap;
+        assignFilteredGlobalDescriptorsFromPropertyDescriptorMap(
+            unsafeBlueDescMap,
+            ObjectGetOwnPropertyDescriptors(globalObjectVirtualizationTarget)
+        );
+        for (const key in unsafeBlueDescMap) {
+            if (!(key in defaultGlobalOwnKeysRegistry)) {
+                delete unsafeBlueDescMap[key];
+            }
+        }
+    }
+    env.remap(blueRefs.window, unsafeBlueDescMap);
+    if (endowments) {
+        const filteredEndowments: PropertyDescriptorMap = {};
+        assignFilteredGlobalDescriptorsFromPropertyDescriptorMap(filteredEndowments, endowments);
+        removeWindowDescriptors(filteredEndowments);
+        env.remap(blueRefs.window, filteredEndowments);
+    }
+    // We remap `blueRefs.WindowPropertiesProto` to an empty object because it
+    // is "magical" in that it provides access to elements by id.
+    env.remapProto(blueRefs.WindowProto, ObjectCreate(blueRefs.EventTargetProto));
```

Comment:
This aligns closer with the non-realm approach but still something is missing

Reply:
I believe you're right... but we might be able to pull this off, I have some ideas. For now, this is what we are missing:

```js
Object.getOwnPropertyDescriptors(window.__proto__.__proto__)
{Symbol(Symbol.toStringTag): {…}}
Symbol(Symbol.toStringTag): {value: 'WindowProperties', writable: false, enumerable: false, configurable: true}
[[Prototype]]: Object
```

This is missing the symbol.toStringTag value, nothing else.

Ideas: map to a proxy of `window.__proto__.__proto__` that hides indexes that should be not leaked.

Reply:
> This is missing the symbol.toStringTag value, nothing else.
>
> Ideas: map to a proxy of window.__proto__.__proto__ that hides indexes that should be not leaked.

Locally I couldn't get any difference in results 😞